connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/ClientDataFrameStatSuite.scala

```diff
@@ -264,4 +264,13 @@ class ClientDataFrameStatSuite extends ConnectFunSuite with RemoteSparkSession {
     }
     assert(error3.getCondition === "DATATYPE_MISMATCH.VALUE_OUT_OF_RANGE")
   }
+
+  test("SPARK-49961: transform type should be consistent") {
+    val session = spark
+    import session.implicits._
+    val ds = Seq(1, 2).toDS()
+    val f: Dataset[Int] => Dataset[Int] = d => d.selectExpr("(value + 1) value").as[Int]
+    val transformed = ds.transform(f)
+    assert(transformed.collect().sorted === Array(2, 3))
```

Comment:
This would normally fail with a ClassCastException right?

Reply:
it shouldn't have a runtime issue as your original impl was base class => base class casting, in fact I'm not sure why the cast was there.  Most of the errors are compilation related. There are some "gaps" (per last comment of mine) with this.type and java.